OpenRA.Mods.Common/Effects/LineImpactProjectile.cs

```diff
@@ -0,0 +1,258 @@
+		{
+			headTicks++;
+			if (headTicks >= length)
+			{
+				headPos = target;
+				headTravelling = false;
+			}
+			else if (headTravelling)
+				headPos = WPos.LerpQuadratic(args.Source, target, angle, headTicks, length);
+
+			if (tailTicks <= 0 &&
+				args.SourceActor.IsInWorld && !args.SourceActor.IsDead)
+			{
+				args.Source = args.SourceActor.CenterPosition + offSet;
+				tailPos = args.Source;
+			}
+
+			// Allow for 1 cell (1024) leniency, to avoid edge case stuttering (start firing and immediately stop again).
+			var outOfWeaponRange = this.args.Weapon.Range.Length + 1024 < (args.PassiveTarget - args.Source).Length;
+
+			// While the head is travelling, the tail must start to follow BeamLengthDuration ticks later.
+			// Alternatively, also stop emitting the beam if source actor dies or is ordered to stop.
+			if ((headTicks >= info.BeamFireDuration && !tailTravelling) || args.SourceActor.IsDead || !args.SourceActor.Trait<AttackBase>().IsAttacking
+				|| outOfWeaponRange)
+				tailTravelling = true;
+
+			if (tailTravelling)
+			{
+				tailTicks++;
+				if (tailTicks >= length)
+				{
+					tailPos = target;
+					tailTravelling = false;
+				}
+				else
+					tailPos = WPos.LerpQuadratic(args.Source, target, angle, tailTicks, length);
+			}
+
+			// Raycast to find all affected actors
+			var hitActors = world.FindActorsOnLine(tailPos, headPos, info.BeamWidth, info.TargetExtraSearchRadius);
+
+			// If this projectile is blockable, first check if any of the hit actors will block this.
+			if (info.Blockable)
+			{
+				var wasBlocked = false;
+				foreach (var victim in hitActors)
+				{
+					if (victim.TraitsImplementing<BlocksProjectiles>().Any(Exts.IsTraitEnabled))
+					{
+						// Find where on the line this actor blocked the beam, and then limit the beam to that point.
+						var projectedPoint = FindActors.MinimumPointLineProjection(tailPos, headPos, victim.CenterPosition);
+						if ((projectedPoint - tailPos).Length < (headPos - tailPos).Length)
+						{
+							headPos = projectedPoint;
+							wasBlocked = true;
+						}
+					}
+				}
+
+				if (wasBlocked)
+				{
+					target = headPos;
+					length = Math.Min(headTicks, length);
+					hitActors = world.FindActorsOnLine(headPos, tailPos, info.BeamWidth, info.TargetExtraSearchRadius);
+				}
+			}
+
+			if (headTicks % info.TicksBetweenImpacts == 0)
+			{
+				foreach (var victim in hitActors)
+				{
+					var currentPos = victim.CenterPosition;
+					var distFromSource = (args.Source - currentPos).Length;
+					var distFromTarget = (target - currentPos).Length;
+
+					// Current falloff multiplier = 100 - ((Current distance - Falloff Start Range) / (Falloff End Range - Falloff Start Range) * Max Falloff);
+					// With the distance limited between 0 and falloff range.
+					var falloffRange = Math.Max((info.FalloffEndDistance - info.FalloffStartDistance).Length, 1);
+					var falloff = (currentPos - args.Source).Length - info.FalloffStartDistance.Length;
```

Comment:
Shouldn't that `(currentPos - args.Source).Length` be the same as `distFromSource`?